### ISSUE TITLE
Problems in calculating the flow around two-dimensional airfoils

### ISSUE THREAD

**Issue by fantasy0645** [lululxvi/deepxde#1556]
Hello Lulu, I am currently using Deepxde to solve a two-dimensional airfoil flow problem. I would like to train a model to predict the pressure coefficient on the airfoil surface, but my results have consistently differed from experimental or CFD values. Below, I am attaching my code and result graph.
I would like to know if there are any areas where my code can be improved, such as sampling or hyperparameters of the network (in this case, I used cfd values as training data, but it is difficult to accurately predict even in input training situations)

```python
import deepxde as dde
import numpy as np
import matplotlib.pyplot as plt
from pyDOE import lhs

dde.config.set_default_float('float64')

global u_inf, rho, L, miu
global data_BL, data_far

V_inf = 27.13
rho = 1.11245
L = 0.9012
miu = 1.7894e-5
R = 20

def load_training_data(num_BL, num_far):

    # 读取数据
    # data_BL = pd.read_csv('2412-2-BL-uvp.dat', delimiter = "\t", header=None)
    # data_far = pd.read_csv('2412-2-far-uvp.dat', delimiter = "\t", header=None)

    #  除去其中的缺失行
    # data_BL = data_BL[data_BL.columns[:5]].dropna()
    # data_far = data_far[data_far.columns[:5]].dropna()

    # 读取数据
    data_BL = np.genfromtxt('bl-uvp.txt', delimiter='\t', dtype = float)
    data_far = np.genfromtxt('far-uvp.txt', delimiter='\t', dtype = float)

    # 边界层网格数据抽取
    idx_BL = np.random.choice(data_BL.shape[0], num_BL, replace=False)
    X_BL_train = data_BL[idx_BL, 0:1]  
    Y_BL_train = data_BL[idx_BL, 1:2]  
    U_BL_train = data_BL[idx_BL, 2:3]
    P_BL_train = data_BL[idx_BL, 3:4]
    V_BL_train = data_BL[idx_BL, 4:5]

    # 流场网格数据抽取
    idx_far = np.random.choice(data_far.shape[0], num_far, replace=False)
    X_far_train = data_far[idx_far, 0:1]  
    Y_far_train = data_far[idx_far, 1:2]  
    U_far_train = data_far[idx_far, 2:3]
    P_far_train = data_far[idx_far, 3:4]
    V_far_train = data_far[idx_far, 4:5]

    # 数据拼接
    x_train = np.concatenate((X_BL_train, X_far_train), axis=0)
    y_train = np.concatenate((Y_BL_train, Y_far_train), axis=0)
    u_train = np.concatenate((U_BL_train, U_far_train), axis=0)
    p_train = np.concatenate((P_BL_train, P_far_train), axis=0)
    v_train = np.concatenate((V_BL_train, V_far_train), axis=0)

    return [x_train, y_train, u_train / V_inf, v_train / V_inf, p_train]

def Navier_Stokes_Equation(x, y):

    Re = rho * V_inf * L / miu

    u = y[:, 0:1]
    v = y[:, 1:2]
    du_x = dde.grad.jacobian(y, x, i=0, j=0)
    du_y = dde.grad.jacobian(y, x, i=0, j=1)
    dv_x = dde.grad.jacobian(y, x, i=1, j=0)
    dv_y = dde.grad.jacobian(y, x, i=1, j=1)
    dp_x = dde.grad.jacobian(y, x, i=2, j=0)
    dp_y = dde.grad.jacobian(y, x, i=2, j=1)
    du_xx = dde.grad.hessian(y, x, component=0, i=0, j=0)
    du_yy = dde.grad.hessian(y, x, component=0, i=1, j=1)
    dv_xx = dde.grad.hessian(y, x, component=1, i=0, j=0)
    dv_yy = dde.grad.hessian(y, x, component=1, i=1, j=1)
    
    continuity = du_x + dv_y
    x_momentum = u * du_x + v * du_y + dp_x - 1 / Re * (du_xx + du_yy)
    y_momentum = u * dv_x + v * dv_y + dp_y - 1 / Re * (dv_xx + dv_yy)

    return [continuity, x_momentum, y_momentum]

# 矩形流场区域范围
x_min, x_max = -1, 2
y_min, y_max = -1, 1

# NACA四位数翼型参数化生成
def boundaryNACA4D(M, P, SS, c, n, offset_x, offset_y, alpha):
    """
    Compute the coordinates of a NACA 4-digits airfoil
    
    Args:
        M:  maximum camber value (*100)
        P:  position of the maximum camber alog the chord (*10)
        SS: maximum thickness (*100)
        c:  chord length
        n:  the total points sampled will be 2*n
    """
    global x_alpha, y_alpha

    m = M / 100
    p = P / 10
    t = SS / 100
    
    if (m == 0):
        p = 1
    
    # Chord discretization (cosine discretization)
    xv = np.linspace(0.0, c, n+1)
    xv = c / 2.0 * (1.0 - np.cos(np.pi * xv / c))
    
    # Thickness distribution
    ytfcn = lambda x: 5 * t * c * (0.2969 * (x / c)**0.5 - 
                                   0.1260 * (x / c) - 
                                   0.3516 * (x / c)**2 + 
                                   0.2843 * (x / c)**3 - 
                                   0.1015 * (x / c)**4)
    yt = ytfcn(xv)
    
    # Camber line
    yc = np.zeros(np.size(xv))
    
    for ii in range(n+1):
        if xv[ii] <= p * c:
            yc[ii] = c * (m / p**2 * (xv[ii] / c) * (2 * p - (xv[ii] / c)))
        else:
            yc[ii] = c * (m / (1 - p)**2 * (1 + (2 * p - (xv[ii] / c)) * (xv[ii] / c) - 2 * p))
    
    # Camber line slope
    dyc = np.zeros(np.size(xv))
    
    for ii in range(n+1):
        if xv[ii] <= p * c:
            dyc[ii] = m / p**2 * 2 * (p - xv[ii] / c)
        else:
            dyc[ii] = m / (1 - p)**2 * 2 * (p - xv[ii] / c)
            
    # Boundary coordinates and sorting        
    th = np.arctan2(dyc, 1)
    xU = xv - yt * np.sin(th)
    yU = yc + yt * np.cos(th)
    xL = xv + yt * np.sin(th)
    yL = yc - yt * np.cos(th)
    
    x = np.zeros(2 * n + 1)
    y = np.zeros(2 * n + 1)
    
    for ii in range(n):
        x[ii] = xL[n - ii]
        y[ii] = yL[n - ii]
        
    x[n : 2 * n + 1] = xU
    y[n : 2 * n + 1] = yU

    x_alpha = (x + offset_x)*np.cos(alpha*np.pi/180)+(y + offset_y)*np.sin(alpha*np.pi/180)
    y_alpha = (y + offset_y)*np.cos(alpha*np.pi/180)-(x + offset_x)*np.sin(alpha*np.pi/180)

    return np.vstack((x_alpha, y_alpha)).T

back = dde.geometry.Rectangle([x_min, y_min], [x_max, y_max])
front = dde.geometry.Disk([1, 0], R)
airfoil = dde.geometry.Polygon(boundaryNACA4D(4, 4, 12, 0.9012, 201, 0.0, 0.0, 0.0))
fluid = dde.geometry.CSGUnion(back, front)
geom = dde.geometry.CSGDifference(fluid, airfoil)

# 获取数据
[ob_x, ob_y, ob_u, ob_v, ob_p] = load_training_data(num_BL = 1000, num_far = 10000) #num_BL = 20451, num_far = 39595

ob_xy = np.hstack((ob_x, ob_y))
observe_u = dde.icbc.PointSetBC(ob_xy, ob_u, component = 0)
observe_v = dde.icbc.PointSetBC(ob_xy, ob_v, component = 1)
observe_p = dde.icbc.PointSetBC(ob_xy, ob_p, component = 2)

# 训练数据整合
data = dde.data.PDE(
    geom, 
    Navier_Stokes_Equation,
    [observe_u, observe_v, observe_p],
    # num_boundary = 200,
    # num_domain = 2500,
    anchors = ob_xy,
)

plt.figure(figsize = (32, 18))
plt.scatter(data.train_x_all[:,0], data.train_x_all[:,1], s = 1)
plt.axis('equal')

# 全连接神经网络
layer_size = [2] + [40] * 8 + [3]
activation = "tanh"
initializer = "Glorot uniform"
net = dde.nn.FNN(layer_size, activation, initializer)
resampler = dde.callbacks.PDEPointResampler(period = 100, pde_points = True, bc_points = True)

# 模型的编译、训练和保存
model = dde.Model(data, net)

model.compile("adam", lr = 1e-3)
loss_history, train_state = model.train(iterations = 20000, display_every = 100, callbacks = [resampler])
# loss_history, train_state = model.train(iterations = 10000, display_every = 1)
dde.saveplot(loss_history, train_state, issave = True, isplot = True)

model.compile("L-BFGS-B")
dde.optimizers.config.set_LBFGS_options(
    maxiter = 50000,
)
loss_history, train_state = model.train(model_save_path = "./")
dde.saveplot(loss_history, train_state, issave = True, isplot = True)

f = model.predict(ob_xy, operator = Navier_Stokes_Equation)
print("Mean residual:", np.mean(np.absolute(f)))

# plot
[ob_x, ob_y, ob_u, ob_v, ob_p] = load_training_data(num_BL = 1000, num_far = 10000)
xy_pred = np.hstack((ob_x, ob_y))
# xy_pred = np.squeeze(xy_pred, axis = 2)
uvp_pred = model.predict(xy_pred)
x_pred, y_pred = xy_pred[:, 0], xy_pred[:, 1]
u_pred, v_pred, p_pred = uvp_pred[:, 0], uvp_pred[:, 1], uvp_pred[:, 2]
# x_true = np.squeeze(ob_x[:, 0], axis = 1)
# y_true = np.squeeze(ob_y[:, 0], axis = 1)
# u_true = np.squeeze(ob_u[:, 0], axis = 1)
# v_true = np.squeeze(ob_v[:, 0], axis = 1)
# p_true = np.squeeze(ob_p[:, 0], axis = 1)
x_true = np.squeeze(ob_x[:, 0])
y_true = np.squeeze(ob_y[:, 0])
u_true = np.squeeze(ob_u[:, 0])
v_true = np.squeeze(ob_v[:, 0])
p_true = np.squeeze(ob_p[:, 0])

airfoil_plot = boundaryNACA4D(4, 4, 12, 0.9012, 201, 0.0, 0.0, 0.0)

fig, ax = plt.subplots(2, 3, figsize = (30, 10))
cntr0 = ax[0, 0].tricontourf(x_pred, y_pred, u_pred, levels = 80, cmap = "rainbow")
cb0 = plt.colorbar(cntr0, ax = ax[0])
cntr1 = ax[1, 0].tricontourf(x_true, y_true, u_true, levels = 80, cmap = "rainbow")
cb1 = plt.colorbar(cntr1, ax = ax[1])
# cntr2 = ax[2, 0].tricontourf(x_true, y_true, u_pred - u_true, levels = 80, cmap = "rainbow")
# cb2 = plt.colorbar(cntr1, ax = ax[1])

cntr3 = ax[0, 1].tricontourf(x_pred, y_pred, v_pred, levels = 80, cmap = "rainbow")
cb3 = plt.colorbar(cntr0, ax = ax[0])
cntr4 = ax[1, 1].tricontourf(x_true, y_true, v_true, levels = 80, cmap = "rainbow")
cb4 = plt.colorbar(cntr1, ax = ax[1])
# cntr5 = ax[2, 1].tricontourf(x_true, y_true, v_pred - v_true, levels = 80, cmap = "rainbow")
# cb5 = plt.colorbar(cntr1, ax = ax[1])

cntr6 = ax[0, 2].tricontourf(x_pred, y_pred, p_pred, levels = 80, cmap = "rainbow")
cb6 = plt.colorbar(cntr0, ax = ax[0])
cntr7 = ax[1, 2].tricontourf(x_true, y_true, p_true, levels = 80, cmap = "rainbow")
cb7 = plt.colorbar(cntr1, ax = ax[1])
# cntr8 = ax[2, 2].tricontourf(x_true, y_true, p_pred - p_true, levels = 80, cmap = "rainbow")
# cb8 = plt.colorbar(cntr1, ax = ax[1])


for i in range(0, 2) :
    for j in range(0, 3) :
        ax[i, j].fill(x_alpha, y_alpha)
        ax[i, j].set_xlim([-1, 2])
        ax[i, j].set_ylim([-1.0, 1.0])

airfoil_xy = np.vstack((x_alpha, y_alpha))
cp_cfd = np.genfromtxt('4412cp-cfd.txt', delimiter='\t', dtype = float)
cp_exp = np.genfromtxt('cp-exp.txt', delimiter='\t', dtype = float)
uvp_wall = model.predict(airfoil_xy.T)
p_wall = uvp_wall[:, 2]
np.savetxt('p_wall.txt', p_wall)
plt.plot(x_alpha, p_wall)
plt.scatter(cp_cfd[:, 0], cp_cfd[:, 1], c = 'r', marker = '+')
plt.scatter(cp_exp[:, 0], cp_exp[:, 1], c = 'r', marker = 'o')
```

loss
![2CUOZ9V{)JRJHFH_JX5WF`C](https://github.com/lululxvi/deepxde/assets/85955827/2dfe6bee-1deb-4a48-a3b0-67999f752a7e)

sample result 
![sample](https://github.com/lululxvi/deepxde/assets/85955827/6bb118c7-759f-45ad-a8af-9576836cfb05)

pressure coefficient(the blue line is the prediction value)
![cp](https://github.com/lululxvi/deepxde/assets/85955827/de7857d2-44d8-49a1-a047-79464d5551b3)

**Comment by llxlnx** [comment]
请问如果不加CFD训练数据，计算结果咋样

**Comment by xmuyao11** [comment]
比加数据集训练结果差，如果能训练到10-3这个数量级误差就没那么大

---- Replied Message ----
| From | ***@***.***> |
| Date | 01/07/2024 00:51 |
| To | lululxvi/deepxde ***@***.***> |
| Cc | xmuyao11 ***@***.***>,
Comment ***@***.***> |
| Subject | Re: [lululxvi/deepxde] Problems in calculating the flow around two-dimensional airfoils (Issue #1556) |

请问如果不加CFD训练数据，计算结果咋样

**Comment by llxlnx** [comment]
请问如果不加训练集，压力系数的预测误差有多大呢

**Comment by xmuyao11** [comment]
压力的误差会比较大，换成压力梯度会好很多

---- Replied Message ----
| From | ***@***.***> |
| Date | 01/17/2024 22:10 |
| To | lululxvi/deepxde ***@***.***> |
| Cc | xmuyao11 ***@***.***>,
Comment ***@***.***> |
| Subject | Re: [lululxvi/deepxde] Problems in calculating the flow around two-dimensional airfoils (Issue #1556) |

请问如果不加训练集，压力系数的预测误差有多大呢
